actors/verifreg/src/lib.rs

```diff
@@ -908,89 +946,116 @@ fn validate_tokens_received(
+    if req.term_max > policy.maximum_verified_allocation_term {
+        return Err(actor_error!(
+            illegal_argument,
+            "allocation term max {} above limit {}",
+            req.term_max,
+            policy.maximum_verified_allocation_term
+        ));
+    }
+    // Term range must be non-empty.
+    if req.term_min > req.term_max {
+        return Err(actor_error!(
+            illegal_argument,
+            "allocation term min {} exceeds term max {}",
+            req.term_min,
+            req.term_max
+        ));
     }
 
-    // Allocation size must match the tokens received exactly (we can't return change).
-    if &allocation_total != datacap_amount {
+    // Allocation must expire in the future.
+    if req.expiration < curr_epoch {
+        return Err(actor_error!(
+            illegal_argument,
+            "allocation expiration epoch {} has passed current epoch {}",
+            req.expiration,
+            curr_epoch
+        ));
+    }
+    // Allocation must expire soon enough.
+    let max_expiration = curr_epoch + policy.maximum_verified_allocation_expiration;
+    if req.expiration > max_expiration {
         return Err(actor_error!(
             illegal_argument,
-            "total allocation size {} must match data cap amount received {}",
-            allocation_total,
-            datacap_amount
+            "allocation expiration {} exceeds maximum {}",
+            req.expiration,
+            max_expiration
         ));
     }
+    Ok(())
+}
 
-    Ok(reqs)
+fn validate_claim_extension(
+    req: &ClaimExtensionRequest,
+    claim: &Claim,
+    policy: &Policy,
+    curr_epoch: ChainEpoch,
+) -> Result<(), ActorError> {
+    // The new term max is the policy limit after current epoch (not after the old term max).
+    let term_limit_absolute = curr_epoch + policy.maximum_verified_allocation_term;
+    let term_limit_relative = term_limit_absolute - claim.term_start;
+    if req.term_max > term_limit_relative {
+        return Err(actor_error!(
+            illegal_argument,
+            format!(
+                "term_max {} for claim {} exceeds maximum {} at current epoch {}",
+                req.term_max, req.claim, term_limit_relative, curr_epoch
+            )
+        ));
+    }
+    // The new term max must be larger than the old one.
+    // Cannot reduce term, and cannot spend datacap on a zero increase.
+    // There is no policy on minimum extension duration.
+    if req.term_max <= claim.term_max {
+        return Err(actor_error!(
+            illegal_argument,
+            "term_max {} for claim {} is not larger than existing term max {}",
+            req.term_max,
+            req.claim,
+            claim.term_max
+        ));
+    }
+    // The claim must not have already expired.
+    // Unlike when the claim client extends term up to the originally-allowed max,
+    // allowing extension of expired claims with new datacap could revive a claim arbitrarily
```

Comment:
checking for understanding: a claim can be extended continuously arbitrarily far into the future but once it has expired it is expired for good?

Reply:
Correct